tests/generation/test_utils.py

```diff
@@ -459,7 +459,8 @@ def test_greedy_generate_dict_outputs(self):
                 # Retrocompatibility check
                 self.assertIsInstance(output_generate, GreedySearchDecoderOnlyOutput)
 
-            self._check_outputs(output_generate, input_ids, model.config)
+            config = model.config if not hasattr(model.config, "text_config") else model.config.text_config
```

Comment:
This feels brittle and makes a strong assumption about the structure of our configs it would be good to have @gante's view on this here. 

I'd suggest moving this handling into `_check_outputs` so the logic change only has to be in on place rather than every `_check_outputs` call, which already has conditional logic on e.g. `config.is_encoder_decoder` within the utility

Reply:
We are starting to have more places where we need to pull the decoder part of a cache in a composite model:
- in the generation config
- in while saving a pretrained config, to ensure it doesn't contain generate flags [open PR]
- here

Note that they are all `generate`-related. In this case, the shape of the `generate` outputs depend on the decoder, so we need to pull the decoder config.

@amyeroberts I was thinking of abstracting this logic into `GenerationMixin`, in a function that pulls the decoder config given a config. WDYT? :)

Reply:
I'd be pro! This should help encapsulate the logic without having to handle it everywhere else.

Reply:
Cool, I'll open a PR after https://github.com/huggingface/transformers/pull/32659 gets merged (to avoid conflicts :) )

Reply:
Great, will port these to `_check_output` until the PR is merged

Reply:
FYI opening a PR to add the generic function AND fixing `_check_output ` (#33212 also needs it :D )

Reply:
https://github.com/huggingface/transformers/pull/33219